emission/tests/funcTests/TestFunctionTiming.py

```diff
@@ -35,207 +17,114 @@ def setUpClass(self):
-            time.sleep(1)
-            return True
-
-        def function_two():
-            logging.info("Executing function_two")
-            time.sleep(1.5)
-            return True
-
-        functions = [function_one, function_two]
+        with ect.Timer() as timer:
+            sample_function()
 
-        with test_logger("test_multiple_functions_timing"):
-            for func in functions:
-                self.run_with_timing(func.__name__, func)
+        esdsq.store_dashboard_time("sample_function", timer)
 
+        # verify that an entry was stored correctly
+        self.verify_stats_entries([
+            ("stats/dashboard_time", "sample_function", timer.elapsed_ms)
+        ])
 
-    def test_code_block_instrumentation(self):
+    def test_multiple_functions_timing(self):
         """
-        Test instrumentation of a specific code block using the Timer context manager.
-        Ensures that timing information for the code block is recorded correctly.
+        Test execution and timing of 2 functions, where we record and validate
+        the time taken for: (i) function_one, (ii) function_two, and (iii) both functions together.
         """
-        block_name = "sample_code_block"
+        def function_one(): return time.sleep(1)
+        def function_two(): return time.sleep(1.5)
 
-        def code_block():
-            logging.info("Executing sample_code_block with simulated operations")
-            logging.info("Step 1: Data retrieval from mock service")
-            time.sleep(0.5)  # Simulate network delay
-            data = {"user_id": 123, "action": "login"}
-            logging.info(f"Data retrieved: {data}")
+        with ect.Timer() as timer_both:
+            with ect.Timer() as timer_one:
+                function_one()
+            esdsq.store_dashboard_time("function_one", timer_one)
 
-            logging.info("Step 2: Data processing")
-            time.sleep(1)  # Simulate processing time
-            processed_data = {k: str(v).upper() for k, v in data.items()}
-            logging.info(f"Processed data: {processed_data}")
+            with ect.Timer() as timer_two:
+                function_two()
+            esdsq.store_dashboard_time("function_two", timer_two)
 
-            logging.info("Step 3: Storing processed data")
-            time.sleep(0.5)  # Simulate storage delay
-            logging.info("Processed data stored successfully")
+        esdsq.store_dashboard_time("functions_one_and_two", timer_both)
 
-        with test_logger("test_code_block_instrumentation"):
-            self.run_with_timing(block_name, code_block)
+        # function_one should've taken ~1000ms; function_two should've taken ~1500ms;
+        # both functions together should've taken ~2500ms
+        self.assertAlmostEqual(timer_one.elapsed_ms, 1000, delta=100)
+        self.assertAlmostEqual(timer_two.elapsed_ms, 1500, delta=100)
+        self.assertAlmostEqual(timer_both.elapsed_ms, 2500, delta=100)
```

Comment:
Below, we will call `verify_stats_entries`, which will make sure that the recorded entries line up with what the timers recorded.

But I also wanted to verify that what the timers recorded is actually reasonable. `function_one` sleeps for 1s so it should finish in a little over 1000ms. I allowed 100ms tolerance by using `assertAlmostEqual` with `delta=100`. This was convenient because it saves us from writing out our own `if` / `else` blocks to handle tolerance

Note that I didn't specifically know that `delta` existed. I also couldn't remember the name for `assertAlmostEqual`. But I googled "unittest close to" which led me to documentation for `assertAlmostEqual` and I read about the `delta` parameter there.
My point here is that you don't need to know everything, but if you learn where to find things you will save time and write cleaner code

Reply:
I thought we didnt need tolerance?